File: VIF_git/GSDS-Qual/algorithms/dynamic-programming/64-minimum-path-sum/minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> mem(m, vector<int>(n, INT_MAX));

        return dp(m-1, n-1, grid, mem);
    }

    int dp(int i, int j, vector<vector<int>>& g, vector<vector<int>>& mem){
        if (i < 0 || j < 0){
            return INT_MAX;
        }
        if (i == 0 && j == 0){
            return g[i][j];
        }

        if (mem[i][j] != INT_MAX){
            return mem[i][j];
        }

        int result = min(dp(i-1, j, g, mem), dp(i, j-1, g, mem)) + g[i][j];
        mem[i][j] = result;
        return result;
    }
};

File: VIF_git/GSDS-Qual/algorithms/dynamic-programming/64-minimum-path-sum/minimum-path-sum.py
from functools import cache
from typing import List


class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        import math

        m, n = len(grid), len(grid[0])

        @cache
        def dp(x: int, y: int) -> int | float:
            if not (0 <= x < m and 0 <= y < n):
                return math.inf
            if x == m - 1 and y == n - 1:
                return grid[x][y]

            return min(dp(x + 1, y), dp(x, y + 1)) + grid[x][y]

        return dp(0, 0)
